Move scheduler basics documentation to Airflow docs

yesemsanthoshkumar opened the pull request:
Resolves #10185 

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

yesemsanthoshkumar committed:
apache#10185 Create first draft of scheduler docs

yesemsanthoshkumar committed:
apache#10185 Add flow diagram for scheduler

mik-laj commented on docs/scheduler.rst:
```diff
@@ -18,11 +18,41 @@
+
+2. ``DAGFileProcessorManager`` - Responsible for listing the files in the DagBag and creating new DAGFileProcessors when required
+
+3. ``SchedulerJob`` - Responsible for sending the TaskInstances to the executor
```
In the Scheduler process, we also have an executor.

yesemsanthoshkumar replied:
AFAIK, the scheduerJob is responsible for sending TI from one state to another and run them. Would `Responsible for transitioning the task instances from one state to another and executing them via a configured executor` be a better explanation here? Please correct me if I'm wrong.

yesemsanthoshkumar replied:
@mik-laj Updated the descriptions. Please take a look.

mik-laj commented on docs/scheduler.rst:
```diff
@@ -18,11 +18,41 @@
+
+1. Airflow kicks off DAGFileProcessorManager along with the SchedulerJob. The DAGFileProcessorManager enumerates all the files in the DAG directory.
+
+2. The DAGFileProcessorManager spawns child processes known as DAGFileProcessor which is responsible for creating the necessary DAG Runs and TaskInstances. When it determines that task instances should run, it updates their state to ``SCHEDULED``. The number of DAGFileProcessor to spawn is configurable via ``num_runs`` in ``airflow.cfg``. TODO: Check if the configuration is correct
```
> The number of DAGFileProcessor to spawn is configurable via ``num_runs`` in ``airflow.cfg``. TODO: Check if the configuration is correct

That's not true. num_run specifies the number of scheduler loop runs.  The maximum number of DAGFileProcesses is determined by the max_threads options in the scheduler section.

yesemsanthoshkumar replied:
So num_runs is about how many iterations the scheduler go over the DagBag, after which the scheduler shut itself down gracefully?

mik-laj commented on docs/scheduler.rst:
```diff
@@ -18,11 +18,41 @@
+
+2. The DAGFileProcessorManager spawns child processes known as DAGFileProcessor which is responsible for creating the necessary DAG Runs and TaskInstances. When it determines that task instances should run, it updates their state to ``SCHEDULED``. The number of DAGFileProcessor to spawn is configurable via ``num_runs`` in ``airflow.cfg``. TODO: Check if the configuration is correct
+
+3. If any of the DAGFileProcessor have finished, another process is spawned to work on the next file in the series.
```
```suggestion
3. If any of the DAGFileProcessor have finished, another process is created to work on the next file in the series.
```
Depending on the operating system, new processes can be spawn(MacOS) or fork(Linux).

mik-laj commented on docs/scheduler.rst:
```diff
@@ -18,11 +18,41 @@
+If the task instance has a state of NONE it will be set to SCHEDULED if the scheduler determines that it needs to run.
+Tasks in the SCHEDULED state are sent to the executor, at which point it is put into the QUEUED state until it actually runs. TODO: Add links to TaskInstance in the documentation. The reason that the task instances are created in the ``SCHEDULED`` state, but then are set to the ``QUEUED`` state once it is sent to the executor, is to ensure that a task instance isn't repeatedly send to the executo ...
+
+Since the scheduler can run indefinitely, it's necessary to periodically refresh the list of files in the DAG definition directory. The refresh interval is controlled with the ``dag_dir_list_interval`` configuration parameter. In cases where there are only a small number of DAG definition files, the loop could potentially process the DAG definition files many times a minute. To control the rate o ...
```
```suggestion
Since the scheduler can run indefinitely, it's necessary to periodically refresh the list of files in the DAG definition directory. The refresh interval is controlled with the ``dag_dir_list_interval`` option in ``[scheduler]`` section. In cases where there are only a small number of DAG definition files, the loop could potentially process the DAG definition files many times a minute. To control the rate of DAG file processing, the ``min_file_process_interval`` option in ``[scheduler]`` section can be set to a higher value. This parameter ensures that a DAG definition file is not processed more often than once every ``min_file_process_interval`` seconds.
```

yesemsanthoshkumar committed:
apache#10185 Reorganise scheduler info

Some of the scheduler related information is best fit into contributors
while some of them has to be for the end user. This commit will
reorganize contents as such.

yesemsanthoshkumar committed:
apache#10185 Add sequence diagram for scheduler

yesemsanthoshkumar commented on CONTRIBUTING.rst:
```diff
@@ -645,6 +645,76 @@ snowflake                  slack
+              self.logger.exception(e)
+
+Now it is a bit more messy at the moment. That is due to the fact that DagRuns are not yet first class citizens. In the past Airflow only had a notion of DagRun: it mentioned them in the code but they did not really exist, instead DAGs were instantiated into "Tasks in Time": TaskInstances. This creates architectural issues but also real life ones. Some of the issues in the current scheduler with  ...
+TODO: Add link to roadmap
```
Which section on roadmap should I link this to? Currently linked this to 2.0 planning.

yesemsanthoshkumar committed:
apache#10185 Remove unwanted todo

yesemsanthoshkumar committed:
apache#10185 Move scheduler images into docs/img
